### ISSUE TITLE
No warning if missing a @required parameter in an overridden method

### ISSUE THREAD

**Issue by kinex** [dart-lang/sdk#36108]
I would expect that I have to declare `@required` parameters only in the base class and not to repeat them in a derived class when overriding a method. Or optionally do not allow an overridden method signature to be different.

```Dart
import 'package:meta/meta.dart';

void main() {
  final A b = B();
  // this results a warning "The paremeter 'i' is required" (correct)
  b.func();

  final B b2 = B();
  // this DOES NOT give a warning as it should
  b2.func();

  final b3 = B();
  // this DOES NOT give a warning as it should
  b3.func();
}

abstract class A {
  void func({@required int i});
}

class B implements A {
  @override
  void func({int i}) {
  }
}

```

Environment:
Dart 2.1.2 (build 2.1.2-dev.0.0 0a7dcf17eb)
Windows/VS Code

**Comment by lrhn** [comment]
It makes sense to allow a subclass to make a required superclass argument optional.
Example:
```dart
class Super {
  void doIt(int number) { ... }
}
class Sub extends Super {
  @override
  void doit([int number = 42]) { ... }
}
```
This is a valid override in plain Dart, it makes a positional required super-class method parameter into an optional parameter in the subclass.
So, when `@required` is used to mark a named parameter as required, it also makes sense to allow a subclass to make that parameter optional, which is what omitting the `@required` means.
Automatically inheriting the `@required` would prevent subclasses from doing so.

**Comment by kinex** [comment]
Yes I agree on that. But I think my example is slightly different because the super-class method is abstract (it does not contain a method body). 

Does it make sense that in an abstract method declaration a parameter is declared as required but then you would allow making it optional in the actual implementation of the method? In my opinion that does not make sense.

I have two (optional) improvement ideas:
1. If an abstract method has a `@required` parameter, derive it automatically to the method implementation. Do this only for abstract methods and only if it is the first override (in a class tree) of an abstract method.

2. Show a warning if an overridden method signature does not match exactly the signature of the super-class abstract method.

**Comment by eernstg** [comment]
Or introduce something like `@unoverridablyRequired`. ;)

**Comment by srawlins** [comment]
The `@required` annotation is being replaced by the `required` keyword, available in Dart 2.12, in the Null safety language feature.
